docs/design/mp_pod_to_vpc_subnet.md

```diff
@@ -0,0 +1,126 @@
+<!--
+SPDX-License-Identifier: MIT
+Copyright (c) 2021 The Authors.
+
+Authors: Phu Tran          <@phudtran>
+
+-->
+
+# Mizar User defined VPCs and Subnets for Vanilla Kubernetes
+
+## Introduction
+
+This document defines the interface for users to define their own VPCs and subnets. Furthermore, it also describes the interface for assigning pods to these specific subnets and VPCs.
+
+## Custom VPC Creation
+
+When creating a VPC, users must define the VPC's name, IP, prefix, and the number of dividers.
+
+An example is given below.
+
+```yaml
+apiVersion: "mizar.com/v1"
+kind: Vpc
+metadata:
+  name: vpc0
+spec:
+  ip: "12.0.0.0"
+  prefix: "8"
+  dividers: 2
+```
+
+#### Requirements
+- The CIDR range is valid.
+- The VPC must not already exist.
+
+## Custom Subnet Creation
+
+When creating a subnet users must define the subnet's name, IP, prefix  the number of bouncers, and the VPC that it belongs to.
+
+An example is given below.
+
+```yaml
+apiVersion: "mizar.com/v1"
+kind: Subnet
+metadata:
+  name: net0
+spec:
+  vpc: "vpc1"
+  ip: "12.2.0.0"
+  prefix: "16"
+  bouncers: 2
+```
+#### Requirements
+- The CIDR range the subnet is given must be within the range of the parent VPC's range.
+- The Parent VPC must exist.
+- The new subnet must not yet already exist within the VPC.
+
+## Pod to VPC and Subnet designation
+
+Three options are defined for assigning pods to specific VPCs.
+
+### Placing a pod into the default VPC
+
+Creating a standard pod with no Mizar specific annotation will assign it to the default VPC that is created at Mizar setup.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+spec:
+  containers:
+  - name: nginx
+    image: nginx
+    ports:
+      - containerPort: 443
+```
+
+### Placing a pod into a specific VPC
+
+Users can place a pod into a specific VPC by specifying the VPC annotation like below.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  annotations:
+    mizar.com: {"vpc": "vpc0"}
+spec:
+  containers:
+  - name: nginx
+    image: nginx
+    ports:
+      - containerPort: 443
+```
+
+#### Requirements
+- The Parent VPC must exist.
+- The VPC must have an existing subnet.
+
+The pod is invalid if the VPC does not exist.
+
+### Placing a pod into a specific VPC and Subnet
+
+Users can place a pod into a specific VPC and subnet by specifying the required VPC and subnet annotation like below.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  annotations:
+    mizar.com: {"vpc": "vpc0", "subnet": "subnet0"}
```

Comment:
Inside XDP programs (for example, the transit XDP), I think VPCs and subnets are referenced by VNI instead of names. In the  In the edge cluster situation, there will be cases where two subnets from two different clusters  belong to the same VPC. When a packet travels from one cluster to the other, VNI will be used in the process for matching, and if the VPC on one cluster has a different VNI from the VPC on the other cluster, it could drop the packet even though the VPCs on both clusters have the same name. In fact, there could also be a case where two VPCs in two cluster have the same name but with two different VNIs. They are different VPC in essence. 

In summary, the VPC name -> VNI mapping works correctly in one cluster, but cross-cluster we will answer what is the unique identifier of VPC? For example, if two VPC has the same name but different VNI, is that considered one VPC, what about different names but the same VNI?

This needs to be answered or it will lead to complication down the road. I don't have an answer yet.

Reply:
I'd tent to think VPC name is the unique identifier cross-clusters. The different VNI could be converted by the edge gateway.

Reply:
Discussed with @vinaykul  and @phudtran offline about this. The conclusion at the moment is to have the VPC name as the unique identifier of VPC. This means within a cluster, Mizar will generate VNI (VPC id) locally, and Edge Gateway will perform translation of VNI in and out of its cluster based on VPC name.

Reply:
Sounds good. We'll merge this doc and work on this design. If we need to change something, we can always revisit.